File: 1-ADT/punto2D/punto2D.js
/*
  Implementacion del ADT Punto2D

  @author: Jorge Londono

  Para ejecutar: node punto2D.js

 */



// Puntos 2D en representacion cartesiana
function Punto2DCartesiano(x,y) {
  this.x = x;
  this.y = y;
  this.toString = function() { return "("+this.x+","+this.y+")" };
  this.getX = function() { return this.x };
  this.getY = function() { return this.y };
}

// Puntos 2D en representacion polar
function Punto2DPolar(r,t) {
  this.toRadians = function(t) { return t*Math.PI/180; };
  this.toDegrees = function(t) { return t*180/Math.PI; };
  this.r = r;
  this.theta = this.toRadians(t);
  this.toString = function() { return `Magnitud: ${this.r}, Angulo: ${this.toDegrees(this.theta)}` };
  this.getX = function() { return this.r*Math.cos(this.theta) };
  this.getY = function() { return this.r*Math.sin(this.theta) };
}

const distancia = function(p2) {
  return Math.sqrt(
    (this.getX()-p2.getX())*(this.getX()-p2.getX()) + (this.getY()-p2.getY())*(this.getY()-p2.getY())
  ) };

const equals = function(p2) {
  return this.getX()===p2.getX() && this.getY()===p2.getY();
}

// Utilizar una misma implementacion del metodo distancia para ambas clases
Punto2DCartesiano.prototype.distancia = distancia;
Punto2DPolar.prototype.distancia = distancia;

// Utilizar una misma implementacion del metodo equals para ambas clases
Punto2DCartesiano.prototype.equals = equals;
Punto2DPolar.prototype.equals = equals;


// Ejemplos de uso de los objetos

let punto1 = new Punto2DCartesiano(5,4);
console.log(punto1.toString());

let punto2 = new Punto2DPolar(7,30);
console.log(punto2.toString());
console.log("("+punto2.getX()+","+punto2.getY()+")");

// Observar el uso polimorfico de los métodos distancia e equals
console.log( punto1.distancia(punto2) );
console.log( punto2.distancia(punto1) );
console.log( punto1.equals(punto2) );

// Son iguales punto3 y punto4? Analizar
let punto3 = new Punto2DCartesiano(1,1);
let punto4 = new Punto2DPolar(Math.sqrt(2), 45);
console.log( punto3.equals(punto4) );

File: 1-ADT/punto2D/Punto2D.rs

trait Punto2D {
    fn get_x(&self) -> f64;
    fn get_y(&self) -> f64;
    fn distancia(&self, p: &dyn Punto2D) -> f64;
    fn to_string(&self) -> String;
}

struct Punto2DCartesiano {
    x: f64,
    y: f64
}

struct Punto2DPolar {
    radio: f64,
    theta: f64
}

impl Punto2D for Punto2DCartesiano {
    fn get_x(&self) -> f64 {
        self.x
    }
    fn get_y(&self) -> f64 {
        self.y
    }
    fn distancia(&self, p: &dyn Punto2D) -> f64 {
        ((self.x-p.get_x())*(self.x-p.get_x()) + (self.y-p.get_y())*(self.y-p.get_y())).sqrt()
    }
    fn to_string(&self) -> String {
        format!("Punto cartesiano: ({},{})", self.x, self.y)
    }
}


impl Punto2D for Punto2DPolar {
    fn get_x(&self) -> f64 {
        self.radio*self.theta.cos()
    }
    fn get_y(&self) -> f64 {
        self.radio*self.theta.sin()
    }
    fn distancia(&self, p: &dyn Punto2D) -> f64 {
        ((self.get_x()-p.get_x())*(self.get_x()-p.get_x()) + (self.get_y()-p.get_y())*(self.get_y()-p.get_y())).sqrt()
    }
    fn to_string(&self) -> String {
        format!("Punto polar: radio={}, angulo={})", self.radio, self.theta)
    }
}


impl Punto2DCartesiano {
    fn new(x: f64, y: f64) -> Punto2DCartesiano {
        return Punto2DCartesiano{ x: x, y: y }
    }
}

impl Punto2DPolar {
    fn new(r: f64, t: f64) -> Punto2DPolar {
        return Punto2DPolar{ radio: r, theta: t }
    }
}


fn main() {
    let a = Punto2DCartesiano::new(5.0, 6.0);
    println!("Punto a: {}", a.to_string());

    let b = Punto2DPolar::new(1.0, std::f64::consts::PI/4.0);
    println!("Punto b: {}", b.to_string());

    let distancia = a.distancia(&b);
    println!("Distancia entre a y b: {}", distancia);

}